### ISSUE TITLE
Add 'job' to Run object

### ISSUE THREAD

**Issue by wslulciuc** [MarquezProject/marquez#1284]
Given a `Run` object, we should be able to easily associate a run with the job (`id`, `name,`, `version`, etc). Let's add `job` to the Run model returned by the API. The `job` field would contain details about the job associated with the run.

**Comment by collado-mike** [comment]
I'm proposing this schema, as shown in the Dataset's `createdByRun` field:

```diff
diff --git a/spec/openapi.yml b/spec/openapi.yml
index 21b72e1e..1af3ffda 100644
--- a/spec/openapi.yml
+++ b/spec/openapi.yml
@@ -945,19 +945,25 @@ components:
           items:
             $ref: '#/components/schemas/DATASET'

+    datasetVersionId:
+      type: object
+      properties:
+        namespace:
+          type: string
+          description: The namespace of the dataset.
+        name:
+          type: string
+          description: The name of the dataset.
+        version:
+          type: string
+          format: uuid
+          description: The version of the dataset
+
     datasetVersion:
       type: object
       properties:
         id:
-          description: The ID of the dataset.
-          type: object
-          properties:
-            namespace:
-              type: string
-              description: The namespace of the dataset.
-            name:
-              type: string
-              description: The name of the dataset.
+          $ref: '#/components/schemas/datasetVersionId'
         type:
           description: The type of the dataset.
           type: string
@@ -1043,6 +1049,16 @@ components:
             args:
               description: The arguments of the run.
               type: object
+            jobVersion:
+              $ref: '#/components/schemas/jobVersionId'
+            inputVersions:
+              type: array
+              items:
+                $ref: '#/components/schemas/datasetVersionId'
+            outputVersions:
+              type: array
+              items:
+                $ref: '#/components/schemas/datasetVersionId'
       example:
         id:
           namespace: my-namespace
@@ -1073,6 +1089,25 @@ components:
        createdByRun: {
          "id": "0e89b60a-d80d-42ae-a12b-bd79ae1fb576",
          "createdAt": "2019-05-09T18:52:35.620763Z",
          "updatedAt": "2019-05-09T18:53:41.461122Z",
          "nominalStartTime": null,
          "nominalEndTime": null,
          "state": "COMPLETED",
          "startedAt": "2021-02-05T18:53:33.245492Z",
          "endedAt": "2021-02-05T18:53:41.461122Z",
          "durationMs": 8216,
          "args": { 'email': 'me@example.com', 'emailOnFailure': false, 'emailOnRetry': true, 'retries': 1 },
+          "jobVersion": {
+            "namespace": "my-namespace",
+            "name": "my_processing_job",
+            "versionUuid": "0ba5b13b-350e-40f3-ba3f-cc00a4e92d11"
+          },
+          "inputVersions": [
+            {
+              "namespace": "my-namespace",
+              "name": "an_input_dataset",
+              "versionUuid": "d224dac0-35d7-4d9b-bbbe-6fff1a8485ad"
+            }
+          ],
+          "outputVersions": [
+            {
+              "namespace": "my-namespace",
+              "name": "my-dataset",
+              "versionUuid": "16e04029-2035-37f1-8206-d57be374333f"
+            }
+          ]
        }
@@ -1309,6 +1354,20 @@ components:
           items:
             $ref: '#/components/schemas/JOB_VERSION'

+    jobVersionId:
+      type: object
+      properties:
+        namespace:
+          type: string
+          description: The namespace of the job.
+        name:
+          type: string
+          description: The name of the job.
+        versionUuid:
+          type: string
+          format: uuid
+          description: The version of the job
+
     JOB_VERSION:
       type: object
       properties:
@@ -1457,6 +1516,16 @@ components:
         args:
           description: The arguments of the run.
           type: object
+        jobVersion:
+          $ref: '#/components/schemas/jobVersionId'
+        inputVersions:
+          type: array
+          items:
+            $ref: '#/components/schemas/datasetVersionId'
+        outputVersions:
+          type: array
+          items:
+            $ref: '#/components/schemas/datasetVersionId'
         facets:
           $ref: '#/components/schemas/runFacets'
```

Each API with a Run object (`latestRun`, `createdByRun`, ...) will get the same modifications. Both the Job and Dataset references in the run will refer to the specific versions.

Also, I think it would be good to establish a convention for the casing of the schema names. Personally, I'm partial to pascal-case (upper case first letter), as it lends itself to code generation. But either way, the mixed upper case and camel case is not preferable.

**Comment by wslulciuc** [comment]
I would use `version` instead of `versionUuid` as we'll want to be consistent with how we name the field. For example, the DatasetVersionAPI uses `version` and when exposing the JobVersionAPI it will use the same name for the field. Otherwise, the spec changes look great!

**Comment by julienledem** [comment]
it would be great if we had a way to make the types Dataset and DatasetVersion more similar. GET /namespaces/{}/datasets/{} is the most recent entry of the array returned by GET /namespaces/{}/datasets/{}/versions

**Comment by wslulciuc** [comment]
> Also, I think it would be good to establish a convention for the casing of the schema names. Personally, I'm partial to pascal-case (upper case first letter), as it lends itself to code generation. But either way, the mixed upper case and camel case is not preferable.

Yeah, I used all uppercase for readability in certain parts of the spec. For example:

<img width="478" alt="Screen Shot 2021-06-22 at 11 42 08 AM" src="https://user-images.githubusercontent.com/1553185/122981684-e2264200-d34e-11eb-9a90-2995e7466a6d.png">

But for your changes to add a job version ID, etc can use camel case. Did you want to open an issue to outline a naming convention for the spec?

**Comment by wslulciuc** [comment]
> it would be great if we had a way to make the types Dataset and DatasetVersion more similar. GET /namespaces/{}/datasets/{} is the most recent entry of the array returned by GET /namespaces/{}/datasets/{}/versions

Yeah, that would be a reasonable underlying change. We currently don't version the dataset description and a dataset has a  `lastModifiedAt` field that we'd need to consider when returning the latest version for a dataset.

**Comment by wslulciuc** [comment]
Fixed #1463
